python/ray/experimental/channel/cpu_nccl_group.py

```diff
@@ -0,0 +1,157 @@
+import asyncio
+from collections import defaultdict
+from typing import Optional, Tuple, List
+from unittest import mock
+
+import torch
+
+import ray
+import ray.experimental.channel as ray_channel
+from ray.experimental.channel.gpu_communicator import TorchTensorAllocator, ReduceOp
+
+
+@ray.remote(num_cpus=0)
+class Barrier:
+    """
+    Barrier that blocks the given number of actors until all actors have
+    reached the barrier. This is used to mock out blocking NCCL ops.
+    """
+
+    def __init__(self, num_actors=2):
+        self.num_actors = num_actors
+        self.condition = asyncio.Condition()
+        # Buffer for the data that is "sent" between the actors, each entry is
+        # one p2p op.
+        self.data = {}
+        self.collective_data = defaultdict(list)
+        # Buffer for the number of actors seen, each entry is one p2p op.
+        self.num_actors_seen = defaultdict(int)
+
+    async def wait(self, op_id: int, data=None):
+        """
+        Wait at barrier until all actors have sent `op_id`. One actor should
+        provide `data`, and this value will be returned by this method for all
+        other actors.
+        """
+        async with self.condition:
+            if data is not None:
+                assert op_id not in self.data, (self.data, self.num_actors_seen)
+                self.data[op_id] = data
+            self.num_actors_seen[op_id] += 1
+
+            if self.num_actors_seen[op_id] == self.num_actors:
+                # Wake up all tasks waiting on this condition.
+                self.condition.notify_all()
+            else:
+                await self.condition.wait_for(
+                    lambda: self.num_actors_seen[op_id] == self.num_actors
+                )
+
+            if data is None:
+                data = self.data[op_id]
```

Comment:
When do we garbage-collect `self.data[op_id]`?

Reply:
We have a strategy for this now, will add comments on it to clarify